File: go/internal/core/paths.go
package core

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strings"
)

// Root is a skills root the producer core can enumerate and export from.
// Disk roots and the embedded bundled root share this abstraction via io/fs.
type Root struct {
	// FS is rooted at the skills root; each immediate subdirectory that
	// contains a SKILL.md is a skill.
	FS fs.FS
	// Key identifies the root for deduplication (absolute path for disk
	// roots, a synthetic marker for embedded roots).
	Key string
	// NormalizeModes forces tar entry modes to 0644/0755 for roots whose
	// backing store loses real file modes (e.g. go:embed).
	NormalizeModes bool
}

// DiskRoot builds a Root backed by the OS filesystem. dir should be absolute.
func DiskRoot(dir string) Root {
	return Root{FS: os.DirFS(dir), Key: dir}
}

// SkillRef is a discovered skill directory inside a Root.
type SkillRef struct {
	ID   string
	Root Root
}

var producerSkillsRoots = []string{"skills", filepath.Join(".agents", "skills")}

// FindSkillsRoots walks upward from start (a file or directory path) looking
// for skills/ then .agents/skills/ directories, returning every root found in
// the first ancestor directory that has any.
func FindSkillsRoots(start string) ([]string, error) {
	current, err := filepath.Abs(start)
	if err != nil {
		return nil, err
	}
	if info, statErr := os.Stat(current); statErr != nil || !info.IsDir() {
		current = filepath.Dir(current)
	}

	for {
		var roots []string
		for _, rel := range producerSkillsRoots {
			candidate := filepath.Join(current, rel)
			if info, statErr := os.Stat(candidate); statErr == nil && info.IsDir() {
				roots = append(roots, candidate)
			}
		}
		if len(roots) > 0 {
			return roots, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			return nil, errors.New(
				"Could not find a skills/ or .agents/skills/ directory. Pass skillsRoot explicitly.",
			)
		}
		current = parent
	}
}

// FindSkillsRoot returns the first root found by FindSkillsRoots.
func FindSkillsRoot(start string) (string, error) {
	roots, err := FindSkillsRoots(start)
	if err != nil {
		return "", err
	}
	return roots[0], nil
}

// AssertValidSkillID validates a user-supplied skill id.
func AssertValidSkillID(id string) error {
	if id == "" || id == "." || id == ".." {
		return errors.New("Skill id is required.")
	}
	if strings.Contains(id, "/") || strings.Contains(id, "\\") {
		return fmt.Errorf("Invalid skill id: %s", id)
	}
	return nil
}

// ListSkillDirs enumerates skill directories directly under a root, sorted
// byte-wise by id. A missing/unreadable root yields no skills.
func ListSkillDirs(root Root) []SkillRef {
	dirents, err := fs.ReadDir(root.FS, ".")
	if err != nil {
		return nil
	}

	var skills []SkillRef
	for _, dirent := range dirents {
		if !dirent.IsDir() {
			continue
		}
		id := dirent.Name()
		if _, statErr := fs.Stat(root.FS, path.Join(id, "SKILL.md")); statErr == nil {
			skills = append(skills, SkillRef{ID: id, Root: root})
		}
	}

	sort.Slice(skills, func(i, j int) bool { return skills[i].ID < skills[j].ID })
	return skills
}

// ResolveSkillFS finds the first root containing the skill and returns a
// filesystem rooted at the skill directory plus whether modes must be
// normalized.
func ResolveSkillFS(roots []Root, id string) (fs.FS, bool, error) {
	if err := AssertValidSkillID(id); err != nil {
		return nil, false, err
	}
	for _, root := range roots {
		if _, err := fs.Stat(root.FS, path.Join(id, "SKILL.md")); err == nil {
			sub, subErr := fs.Sub(root.FS, id)
			if subErr != nil {
				return nil, false, subErr
			}
			return sub, root.NormalizeModes, nil
		}
	}
	return nil, false, fmt.Errorf("Skill not found: %s", id)
}

File: rust/crates/skillflag/src/core/paths.rs
use std::fs;
use std::path::{Component, Path, PathBuf};

use crate::bundled;
use crate::core::errors::{Result, SkillflagError};

/// A skills root: either a directory on disk or the virtual bundled root.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SkillsRoot {
    Dir(PathBuf),
    Bundled,
}

/// A resolved skill directory.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SkillDirRef {
    Disk(PathBuf),
    /// Skill id inside the bundled root.
    Bundled(String),
}

const PRODUCER_SKILLS_ROOTS: [&[&str]; 2] = [&["skills"], &[".agents", "skills"]];

/// Lexically normalize a path: resolve `.` and `..` components without
/// touching the filesystem (mirrors Node's `path.resolve` semantics).
fn normalize(path: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    for component in path.components() {
        match component {
            Component::Prefix(_) | Component::RootDir => out.push(component.as_os_str()),
            Component::CurDir => {}
            Component::ParentDir => {
                out.pop();
            }
            Component::Normal(name) => out.push(name),
        }
    }
    if out.as_os_str().is_empty() {
        PathBuf::from(".")
    } else {
        out
    }
}

/// Resolve `path` against `cwd` (when relative) and normalize lexically.
pub fn resolve_path(cwd: &Path, path: &Path) -> PathBuf {
    if path.is_absolute() {
        normalize(path)
    } else {
        normalize(&cwd.join(path))
    }
}

/// Resolve and dedupe skills roots, preserving order.
pub fn resolve_skills_roots(roots: &[PathBuf], cwd: &Path) -> Vec<PathBuf> {
    let mut resolved: Vec<PathBuf> = Vec::new();
    for root in roots {
        let root = resolve_path(cwd, root);
        if !resolved.contains(&root) {
            resolved.push(root);
        }
    }
    resolved
}

fn existing_producer_roots(dir: &Path) -> Vec<PathBuf> {
    let mut roots = Vec::new();
    for segments in PRODUCER_SKILLS_ROOTS {
        let mut candidate = dir.to_path_buf();
        for segment in segments {
            candidate.push(segment);
        }
        if candidate.is_dir() {
            roots.push(candidate);
        }
    }
    roots
}

/// Walk upward from `start` looking for `skills/` then `.agents/skills/`
/// directories. Returns every root found at the first ancestor that has any.
pub fn find_skills_roots(start: impl AsRef<Path>) -> Result<Vec<PathBuf>> {
    let start = start.as_ref();
    let mut current = std::path::absolute(start).unwrap_or_else(|_| start.to_path_buf());
    let is_dir = fs::metadata(&current).map(|m| m.is_dir()).unwrap_or(false);
    if !is_dir {
        if let Some(parent) = current.parent() {
            current = parent.to_path_buf();
        }
    }

    loop {
        let roots = existing_producer_roots(&current);
        if !roots.is_empty() {
            return Ok(roots);
        }
        match current.parent() {
            Some(parent) if !parent.as_os_str().is_empty() => current = parent.to_path_buf(),
            _ => {
                return Err(SkillflagError::new(
                    "Could not find a skills/ or .agents/skills/ directory. Pass skillsRoot explicitly.",
                ))
            }
        }
    }
}

/// First root returned by [`find_skills_roots`].
pub fn find_skills_root(start: impl AsRef<Path>) -> Result<PathBuf> {
    Ok(find_skills_roots(start)?.remove(0))
}

pub fn assert_valid_skill_id(id: &str) -> Result<()> {
    if id.is_empty() || id == "." || id == ".." {
        return Err(SkillflagError::new("Skill id is required."));
    }
    if id.contains('/') || id.contains('\\') {
        return Err(SkillflagError::new(format!("Invalid skill id: {id}")));
    }
    Ok(())
}

/// List skill directories (id + location) under one root, sorted by id.
pub fn list_skill_dirs(root: &SkillsRoot) -> Vec<(String, SkillDirRef)> {
    match root {
        SkillsRoot::Bundled => bundled::skill_ids()
            .into_iter()
            .map(|id| (id.to_string(), SkillDirRef::Bundled(id.to_string())))
            .collect(),
        SkillsRoot::Dir(dir) => {
            let Ok(read_dir) = fs::read_dir(dir) else {
                return Vec::new();
            };
            let mut skills = Vec::new();
            for entry in read_dir.flatten() {
                let Ok(file_type) = entry.file_type() else {
                    continue;
                };
                if !file_type.is_dir() {
                    continue;
                }
                let Ok(id) = entry.file_name().into_string() else {
                    continue;
                };
                let skill_dir = entry.path();
                if skill_dir.join("SKILL.md").exists() {
                    skills.push((id, SkillDirRef::Disk(skill_dir)));
                }
            }
            skills.sort_by(|a, b| a.0.cmp(&b.0));
            skills
        }
    }
}

/// Resolve a skill id against roots in order; the first root wins.
pub fn resolve_skill_dir_from_roots(roots: &[SkillsRoot], id: &str) -> Result<SkillDirRef> {
    assert_valid_skill_id(id)?;
    for root in roots {
        match root {
            SkillsRoot::Dir(dir) => {
                let skill_dir = dir.join(id);
                if skill_dir.join("SKILL.md").exists() {
                    return Ok(SkillDirRef::Disk(skill_dir));
                }
            }
            SkillsRoot::Bundled => {
                if bundled::has_skill(id) {
                    return Ok(SkillDirRef::Bundled(id.to_string()));
                }
            }
        }
    }
    Err(SkillflagError::new(format!("Skill not found: {id}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_and_invalid_skill_ids() {
        assert!(assert_valid_skill_id("alpha").is_ok());
        for id in ["", ".", ".."] {
            assert_eq!(
                assert_valid_skill_id(id).unwrap_err().message,
                "Skill id is required."
            );
        }
        assert_eq!(
            assert_valid_skill_id("a/b").unwrap_err().message,
            "Invalid skill id: a/b"
        );
        assert_eq!(
            assert_valid_skill_id("a\\b").unwrap_err().message,
            "Invalid skill id: a\\b"
        );
    }

    #[test]
    fn normalize_resolves_dots() {
        assert_eq!(
            resolve_path(Path::new("/base"), Path::new("a/./b/../c")),
            PathBuf::from("/base/a/c")
        );
        assert_eq!(
            resolve_path(Path::new("/base"), Path::new("/abs/./x")),
            PathBuf::from("/abs/x")
        );
    }

    #[test]
    fn bundled_root_lists_bundled_skills() {
        let dirs = list_skill_dirs(&SkillsRoot::Bundled);
        assert_eq!(dirs.len(), 1);
        assert_eq!(dirs[0].0, "skillflag");
    }
}
